File: WebDriver-Series/WebDriver.Series.Tests/GridTestCases/FreightColumnAsserter.cs
using System;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using WebDriver.Series.Tests.Controls;
using WebDriver.Series.Tests.GridTestCases.Pages;

namespace WebDriver.Series.Tests.GridTestCases
{
    public class FreightColumnAsserter : GridColumnAsserter
    {
        public FreightColumnAsserter(IGridPage gridPage) : base(gridPage)
        {
        }

        public void FreightEqualToFilter()
        {
            this.GridPage.NavigateTo();
            
            var newItem = this.CreateNewItemInDb();
            newItem.Freight = this.GetUniqueNumberValue();
            this.UpdateItemInDb(newItem);
            
            this.GridPage.Grid.Filter(GridColumns.Freight, Enums.FilterOperator.EqualTo, newItem.Freight.ToString());
            this.WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();
            
            Assert.IsTrue(results.Count() == 1);
            Assert.AreEqual(newItem.Freight.ToString(), results[0].Freight);
        }
        
        public void FreightGreaterThanOrEqualToFilter()
        {
            this.GridPage.NavigateTo();
            
            var allItems = this.GetAllItemsFromDb().OrderBy(x => x.Freight);
            var biggestFreight = allItems.Last().Freight;
            
            var newItem = this.CreateNewItemInDb();
            newItem.Freight = biggestFreight + this.GetUniqueNumberValue();
            this.UpdateItemInDb(newItem);
            
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            secondNewItem.Freight = newItem.Freight + 1;
            this.UpdateItemInDb(secondNewItem);
            
            this.GridPage.Grid.Filter(GridColumns.Freight, Enums.FilterOperator.GreaterThanOrEqualTo, newItem.Freight.ToString());
            this.WaitForGridToLoadAtLeast(2, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();

            Assert.IsTrue(results.Count() == 2);
            
            // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
            Assert.AreEqual(1, results.Count(x => x.Freight == secondNewItem.Freight));
            Assert.AreEqual(1, results.Count(x => x.Freight == newItem.Freight));
        }
        
        public void FreightGreaterThanFilter()
        {
            this.GridPage.NavigateTo();
            
            var allItems = this.GetAllItemsFromDb().OrderBy(x => x.Freight);
            var biggestFreight = allItems.Last().Freight;
            
            var newItem = this.CreateNewItemInDb();
            newItem.Freight = biggestFreight + this.GetUniqueNumberValue();
            this.UpdateItemInDb(newItem);
            
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            secondNewItem.Freight = newItem.Freight + 1;
            this.UpdateItemInDb(secondNewItem);
            
            this.GridPage.Grid.Filter(GridColumns.Freight, Enums.FilterOperator.GreaterThan, newItem.Freight.ToString());
            this.WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();

            Assert.IsTrue(results.Count() == 1);
            
            // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
            Assert.AreEqual(1, results.Count(x => x.Freight == secondNewItem.Freight));
        }
        
        public void FreightLessThanOrEqualToFilter()
        {
            this.GridPage.NavigateTo();
            
            var allItems = this.GetAllItemsFromDb().OrderBy(x => x.Freight);
            var smallestFreight = allItems.First().Freight;
            
            var newItem = this.CreateNewItemInDb();
            newItem.Freight = Math.Round(smallestFreight - this.GetUniqueNumberValue(), 3, MidpointRounding.AwayFromZero);
            this.UpdateItemInDb(newItem);
            
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            secondNewItem.Freight = Math.Round((newItem.Freight - 0.01), 3, MidpointRounding.AwayFromZero);
            this.UpdateItemInDb(secondNewItem);
            
            this.GridPage.Grid.Filter(GridColumns.Freight, Enums.FilterOperator.LessThanOrEqualTo, newItem.Freight.ToString());
            this.WaitForGridToLoadAtLeast(2, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();

            Assert.IsTrue(results.Count() == 2);
            
            // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
            Assert.AreEqual(1, results.Count(x => x.Freight == secondNewItem.Freight));
            Assert.AreEqual(1, results.Count(x => x.Freight == newItem.Freight));
        }
        
        public void FreightLessThanFilter()
        {
            this.GridPage.NavigateTo();
            
            var allItems = this.GetAllItemsFromDb().OrderBy(x => x.Freight);
            var smallestFreight = allItems.First().Freight;
            
            var newItem = this.CreateNewItemInDb();
            newItem.Freight = Math.Round(smallestFreight - this.GetUniqueNumberValue(), 3, MidpointRounding.AwayFromZero);
            this.UpdateItemInDb(newItem);
            
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            secondNewItem.Freight = Math.Round((newItem.Freight - 0.01), 3, MidpointRounding.AwayFromZero);
            this.UpdateItemInDb(secondNewItem);
            
            this.GridPage.Grid.Filter(GridColumns.Freight, Enums.FilterOperator.LessThan, newItem.Freight.ToString());
            this.WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();

            Assert.IsTrue(results.Count() == 1);
            
            // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
            Assert.AreEqual(1, results.Count(x => x.Freight == secondNewItem.Freight));
        }
        
        public void FreightNotEqualToFilter()
        {
            this.GridPage.NavigateTo();
            
            var newItem = this.CreateNewItemInDb();
            newItem.Freight = this.GetUniqueNumberValue();
            this.UpdateItemInDb(newItem);
            
            // After we apply the orderId filter, only 1 item is displayed in the grid. When we apply the NotEqualTo filter this item will disappear.
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.Freight, Enums.FilterOperator.NotEqualTo, newItem.Freight.ToString()),
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.EqualTo, newItem.OrderId.ToString()));
            this.WaitForGridToLoad(0, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();
        
            Assert.IsTrue(results.Count() == 0);
        }
        
        public void FreightClearFilter()
        {
            this.GridPage.NavigateTo();
            
            var allItems = this.GetAllItemsFromDb().OrderBy(x => x.Freight);
            var biggestFreight = allItems.Last().Freight;
            
            var newItem = this.CreateNewItemInDb();
            newItem.Freight = biggestFreight + this.GetUniqueNumberValue();
            this.UpdateItemInDb(newItem);
            
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            secondNewItem.Freight = newItem.Freight + 1;
            this.UpdateItemInDb(secondNewItem);
            
            this.GridPage.Grid.Filter(GridColumns.Freight, Enums.FilterOperator.EqualTo, newItem.Freight.ToString());
            this.WaitForGridToLoad(1, this.GridPage.Grid);
            this.GridPage.Grid.RemoveFilters();
        
            this.WaitForGridToLoadAtLeast(2, this.GridPage.Grid);
        }
    }
}

File: java/WebAutomation-Series/src/test/java/kendogrid/reuse/FreightColumnAsserter.java
package kendogrid.reuse;

import kendogrid.Order;
import kendogrid.components.FilterOperator;
import kendogrid.components.GridFilter;
import org.openqa.selenium.WebDriver;
import org.testng.Assert;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class FreightColumnAsserter extends GridColumnAsserter {
    public FreightColumnAsserter(IGridPage gridPage, WebDriver driver) {
        super(gridPage, driver);
    }

    public void freightEqualToFilter() throws Exception {
        getGridPage().navigateTo();

        var newItem = createNewItemInDb();
        newItem.setFreight(getUniqueNumberValue());
        updateItemInDb(newItem);

        getGridPage().getGrid().filter(GridColumns.FREIGHT, FilterOperator.EQUAL_TO, Double.toString(newItem.getFreight()));
        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertTrue(results.stream().count() == 1);
        Assert.assertEquals(Double.toString(newItem.getFreight()), results.get(0).getFreight());
    }

    public void freightGreaterThanOrEqualToFilter() throws Exception {
        getGridPage().navigateTo();

        var allItems = getAllItemsFromDb().stream().sorted(Comparator.comparing(Order::getFreight)).collect(Collectors.toList());
        var biggestFreight = allItems.stream().skip(allItems.stream().count() - 1).findFirst().get().getFreight();

        var newItem = createNewItemInDb();
        newItem.setFreight(BigDecimal.valueOf(biggestFreight + getUniqueNumberValue()).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(newItem);

        var secondNewItem = createNewItemInDb(newItem.getShipName());
        secondNewItem.setFreight(BigDecimal.valueOf(newItem.getFreight() + 1).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(secondNewItem);

        getGridPage().getGrid().filter(GridColumns.FREIGHT, FilterOperator.GREATER_THAN_OR_EQUAL_TO, Double.toString(newItem.getFreight()));
        waitForGridToLoadAtLeast(2, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertTrue(results.stream().count() == 2);

        // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
        Assert.assertEquals(results.stream().filter(x -> x.getFreight() == secondNewItem.getFreight()).count(), 1);
        Assert.assertEquals(results.stream().filter(x -> x.getFreight() == newItem.getFreight()).count(), 1);
    }

    public void freightGreaterThanFilter() throws Exception {
        getGridPage().navigateTo();

        var allItems = getAllItemsFromDb().stream().sorted(Comparator.comparing(Order::getFreight)).collect(Collectors.toList());
        var biggestFreight = allItems.stream().skip(allItems.stream().count() - 1).findFirst().get().getFreight();

        var newItem = createNewItemInDb();
        newItem.setFreight(BigDecimal.valueOf(biggestFreight + getUniqueNumberValue()).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(newItem);

        var secondNewItem = createNewItemInDb(newItem.getShipName());
        secondNewItem.setFreight(BigDecimal.valueOf(newItem.getFreight() + 1).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(secondNewItem);

        getGridPage().getGrid().filter(GridColumns.FREIGHT, FilterOperator.GREATER_THAN, Double.toString(newItem.getFreight()));
        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertTrue(results.stream().count() == 1);

        // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
        Assert.assertEquals(results.stream().filter(x -> x.getFreight() == secondNewItem.getFreight()).count(), 1);
    }

    public void freightLessThanOrEqualToFilter() throws Exception {
        getGridPage().navigateTo();

        var allItems = getAllItemsFromDb().stream().sorted(Comparator.comparing(Order::getFreight)).collect(Collectors.toList());
        var smallestFreight = allItems.stream().findFirst().get().getFreight();

        var newItem = createNewItemInDb();
        newItem.setFreight(BigDecimal.valueOf(smallestFreight - getUniqueNumberValue()).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(newItem);

        var secondNewItem = createNewItemInDb(newItem.getShipName());
        secondNewItem.setFreight(BigDecimal.valueOf(newItem.getFreight() - 0.01).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(secondNewItem);

        getGridPage().getGrid().filter(GridColumns.FREIGHT, FilterOperator.LESS_THAN_OR_EQUAL_TO, Double.toString(newItem.getFreight()));
        waitForGridToLoadAtLeast(2, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertTrue(results.stream().count() == 2);

        // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
        Assert.assertEquals(results.stream().filter(x -> x.getFreight() == secondNewItem.getFreight()).count(), 1);
        Assert.assertEquals(results.stream().filter(x -> x.getFreight() == newItem.getFreight()).count(), 1);
    }

    public void freightLessThanFilter() throws Exception {
        getGridPage().navigateTo();

        var allItems = getAllItemsFromDb().stream().sorted(Comparator.comparing(Order::getFreight)).collect(Collectors.toList());
        var smallestFreight = allItems.stream().findFirst().get().getFreight();

        var newItem = createNewItemInDb();
        newItem.setFreight(BigDecimal.valueOf(smallestFreight - getUniqueNumberValue()).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(newItem);

        var secondNewItem = createNewItemInDb(newItem.getShipName());
        secondNewItem.setFreight(BigDecimal.valueOf(newItem.getFreight() - 0.01).setScale(3, RoundingMode.HALF_UP).doubleValue());
        updateItemInDb(secondNewItem);

        getGridPage().getGrid().filter(GridColumns.FREIGHT, FilterOperator.LESS_THAN, Double.toString(newItem.getFreight()));
        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertTrue(results.stream().count() == 1);

        // We assume that there isn't default sorting for this column so we cannot expect that the order will be the same every time.
        Assert.assertEquals(results.stream().filter(x -> x.getFreight() == secondNewItem.getFreight()).count(), 1);
    }

    public void FreightNotEqualToFilter() throws Exception {
        getGridPage().navigateTo();

        var newItem = createNewItemInDb();
        newItem.setFreight(getUniqueNumberValue());
        updateItemInDb(newItem);

        // After we apply the orderId filter, only 1 item is displayed in the grid. When we apply the NotEqualTo filter this item will disappear.
        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.FREIGHT, FilterOperator.NOT_EQUAL_TO, Double.toString(newItem.getFreight())),
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.EQUAL_TO, newItem.getOrderId().toString()));
        waitForGridToLoad(0, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertTrue(results.stream().count() == 0);
    }

    public void freightClearFilter() throws Exception {
        getGridPage().navigateTo();

        var allItems = getAllItemsFromDb().stream().sorted(Comparator.comparing(Order::getFreight)).collect(Collectors.toList());
        var biggestFreight = allItems.stream().skip(allItems.stream().count() - 1).findFirst().get().getFreight();

        var newItem = createNewItemInDb();
        newItem.setFreight(biggestFreight + getUniqueNumberValue());
        updateItemInDb(newItem);

        var secondNewItem = createNewItemInDb(newItem.getShipName());
        secondNewItem.setFreight(newItem.getFreight() + 1);
        updateItemInDb(secondNewItem);

        getGridPage().getGrid().filter(GridColumns.FREIGHT, FilterOperator.EQUAL_TO, Double.toString(newItem.getFreight()));
        waitForGridToLoad(1, getGridPage().getGrid());
        getGridPage().getGrid().removeFilters();

        waitForGridToLoadAtLeast(2, getGridPage().getGrid());
    }
}
